src/nbl/asset/utils/CShaderIntrospector.cpp

```diff
@@ -101,25 +101,31 @@ const CIntrospectionData* CShaderIntrospector::introspect(const ICPUShader* _sha
         return m_introspectionCache[std::move(key)].insert({core::smart_refctd_ptr<const ICPUShader>(_shader),std::move(introspection)}).first->second.get();
     };
 
-    if (_shader->containsGLSL())
+    auto begin = reinterpret_cast<const char*>(_shader->getContent()->getPointer());
+    auto end = begin + _shader->getContent()->getSize();
+    std::string code(begin, end);
+    if (_shader->getContentType() != ICPUShader::E_CONTENT_TYPE::ECT_SPIRV)
+        ICPUShader::insertDefines(code, _params.extraDefines);
+
+    // TODO:
+    core::smart_refctd_ptr<asset::CCompilerSet> compilerSet;
```

Comment:
ask for the compiler set and options  in `SIntrospectionParams`

then patch up any options that don't make sense, like `genDebugInfo`, `optimizer`, `sourceIdentifier`, `includeFinder`

Reply:
the only thing that makes sense is entry point, you want me to make `SIntrospectionParams` hold a CompilerOption and patch it instead of just getting an `entryPoint`? doesn't make sense to me

Reply:
you can make the introspection method take `CCompulerSet*` and `IShaderCompiler::SOptions`

then you can ignore options that get in the way of introspection, such as an `optimizer` or a lack of debug info.

Futhermore you can substitute `sourceIndentier` and `includeFinder` if they're invalid

Reply:
The entry point in `SIntrospectionParams` is also a member in `IShaderCompiler::SOptions`, I think I will remove that but IntrospectionParams will be only this .
```c
struct SIntrospectionParams
{
	// const char* entryPoint;
	// if the Shader is already compiled to SPV, it will ignore this member
	core::SRange<const char* const> extraDefines;
};
```

There is also a Comparator thing which I can't make sense of. https://github.com/Devsh-Graphics-Programming/Nabla/blob/compiler_cleanup/include/nbl/asset/utils/CShaderIntrospector.h#L180
I only know that it's being used to cache introspections

So I need to also account for the CompileOptions in the caching as well. because some language specific compile options might affect how the bindings or resources are positioned, and then It becomes a pain in the ass cause we'll have to cache the `CHLSL/CGLSLCompiler::SOptions` as well, maybe in a union?!

But at the end of the day I think we'll have our compile options inside the `IntrospectionParams`; but the CCompilerSet would be retrieved via function params as follows:

```c
CShaderIntrospector::introspect(const ICPUShader* _shader, CCompilerSet* compilerSet, const SIntrospectionParams& _params)
```

TLDR; I'm mind-f@!ked and this seems non-trivial to fix

Reply:
> There is also a Comparator thing which I can't make sense of. https://github.com/Devsh-Graphics-Programming/Nabla/blob/compiler_cleanup/include/nbl/asset/utils/CShaderIntrospector.h#L180
I only know that it's being used to cache introspections

That's because the SPIR-V produced by a compilation differs on the entry point and the defines you have before.

So if you want to cache introspections for a given `ICPUShader` then you need to make sure that the entry point and defines were identical, that's all.

Reply:
ok, but entry point in the params is duplicated (once in `SIntrospectionParams` and once in IShaderCompiler::Options)
Another thing: this is based on the assumption the compiler options taken in the params doesn't affect the introspection (even if it's patched), otherwise it should be cached